specs/experimental/gov-delegation.md

```diff
@@ -0,0 +1,406 @@
+
+Retrieves the number checkpoints for a given user address.
+
+```solidity
+function numCheckpoints(address _account) external view returns (uint32)
+```
+
+#### `delegates`
+
+Retrieves the delegation of a given user address with the highest voting power.
+This function is intended to be used by the `GovernanceToken` contract to maximize for backwards compatibility.
+
+```solidity
+function delegates(address _account) external view returns (address)
+```
+
+#### `getVotes`
+
+Retrieves the voting power of a given user address.
+
+```solidity
+function getVotes(address _account) external view returns (uint256)
+```
+
+#### `getPastVotes`
+
+Retrieves the voting power of a given user address at a block.
+
+```solidity
+function getPastVotes(address _account, uint256 _blockNumber) external view returns (uint256)
+```
+
+#### `getPastTotalSupply`
+
+Retrieves the total supply of the `GovernanceToken` at a given block.
+
+```solidity
+function getPastTotalSupply(uint256 _blockNumber) external view returns (uint256)
+```
+
+#### `migrated`
+
+Returns the migration status of an account — `True` if the account has been migrated, `False` otherwise.
+
+```solidity
+function migrated(address _account) public view returns (bool)
+```
+
+#### `delegations`
+
+Retrieves the delegations of a given user address sorted in descending order by voting power.
+
+```solidity
+function delegations(address _account) public view returns (Delegation[] memory)
+```
+
+### Events
+
+#### `DelegationCreated`
+
+MUST trigger when an account delegates voting power to one or more addresses (delegatees).
+
+```solidity
+event DelegationCreated(address indexed account, Delegation[] delegations);
+```
+
+#### `DelegateVotesChanged`
+
+MUST trigger every time the voting power of a user changes, including when a token transfer occurs or when a
+delegation is created.
+
+```solidity
+event DelegateVotesChanged(address indexed delegatee, uint256 previousBalance, uint256 newBalance)
+```
+
+## Storage
+
+The `GovernanceDelegation` contract MUST be able to store delegations and checkpoints. These storage variables MUST be
+defined as in the `GovernanceToken` and use the same types:
+
+```solidity
+// Addresses that had their delegation state migrated from the `GovernanceToken` to the `GovernanceDelegation`.
+mapping(address => bool) public migrated;
+
+// Voting power delegations of an account.
+mapping(address => Delegation[]) public delegations;
+
+// Checkpoints of voting power for an account.
+mapping(address => ERC20Votes.Checkpoint[]) internal _checkpoints;
+
+// Total supply checkpoints of the GovernanceToken.
+ERC20Votes.Checkpoint[] internal _totalSupplyCheckpoints;
+```
+
+## Constants
+
+The `GovernanceDelegation` contract MUST define the following constants:
+
+| Name                     | Value     | Description                                              |
+|--------------------------|-----------|----------------------------------------------------------|
```

Comment:
Noticing there are 2 constants tables, one at the top and one here

Reply:
We should combine these two tables

Reply:
you mean combining the storage table and the constants table into one?

Reply:
ohh i think maybe you mean this one in the overview section https://github.com/ethereum-optimism/specs/blob/5997c515116651381b65acc4173b648657b9c296/specs/experimental/gov-delegation.md#overview?

Reply:
Yes better to combine everything into the constants table at the beginning of the doc since that's our current format.

Reply:
updated in https://github.com/ethereum-optimism/specs/pull/247/commits/b7288ea73d8ce7eaa0606d9aba624e0557cd213f